auth/standard_role_manager.cc

```diff
@@ -250,24 +251,32 @@ future<> standard_role_manager::start() {
             }
             return make_ready_future<>();
         }).then([this] {
-            _stopped = auth::do_after_system_ready(_as, [this] () -> future<void>  {
-                if (legacy_mode(_qp)) {
-                    co_await _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as);
+            _stopped = auth::do_after_system_ready(_as, [this] () -> future<>  {
+                auto signal_super_user_created = defer([this] {
+                    _superuser_created.signal();
+                });
+                try {
+                    if (legacy_mode(_qp)) {
+                        co_await _migration_manager.wait_for_schema_agreement(_qp.db().real_database(), db::timeout_clock::time_point::max(), &_as);
+
+                        if (co_await any_nondefault_role_row_satisfies(_qp, &has_can_login)) {
+                            if (legacy_metadata_exists()) {
+                                log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
+                            }
+                            co_return;
+                        }
 
-                    if (co_await any_nondefault_role_row_satisfies(_qp, &has_can_login)) {
                         if (legacy_metadata_exists()) {
-                            log.warn("Ignoring legacy user metadata since nondefault roles already exist.");
+                            co_await migrate_legacy_metadata();
+                            co_return;
                         }
-
-                        co_return;
-                    }
-
-                    if (legacy_metadata_exists()) {
-                        co_await migrate_legacy_metadata();
-                        co_return;
                     }
+                    co_await create_default_role_if_missing();
+                } catch (...) {
+                    // If we fail to create the default role, we should not signal that we are ready.
+                    signal_super_user_created.cancel();
+                    throw;
```

Comment:
So why use `defer` in the first place? It would be equivalent (but much easier to read) to put the `_superuser_created.signal();` directly after the `catch` block.

Reply:
Right.